paasta_tools/autoscaling/autoscaling_cluster_lib.py

```diff
@@ -934,27 +935,42 @@ def autoscale_local_cluster(config_folder, dry_run=False, log_level=None):
     autoscaling_resources = system_config.get_cluster_autoscaling_resources()
     autoscaling_draining_enabled = system_config.get_cluster_autoscaling_draining_enabled()
     all_pool_settings = system_config.get_resource_pool_settings()
-    autoscaling_scalers = []
     mesos_state = get_mesos_master().state
     utilization_errors = get_all_utilization_errors(autoscaling_resources, all_pool_settings, mesos_state)
+    autoscaling_scalers = defaultdict(list)
+    any_cancelled_in_pool = defaultdict(lambda: False)
     for identifier, resource in autoscaling_resources.items():
         pool_settings = all_pool_settings.get(resource['pool'], {})
         try:
-            autoscaling_scalers.append(get_scaler(resource['type'])(
+            scaler = get_scaler(resource['type'])(
                 resource=resource,
                 pool_settings=pool_settings,
                 config_folder=config_folder,
                 dry_run=dry_run,
                 log_level=log_level,
                 utilization_error=utilization_errors[(resource['region'], resource['pool'])],
                 draining_enabled=autoscaling_draining_enabled,
-            ))
+            )
+            autoscaling_scalers[(resource['region'], resource['pool'])].append(scaler)
+            any_cancelled_in_pool[(resource['region'], resource['pool'])] |= scaler.is_resource_cancelled()
         except KeyError:
             log.warning("Couldn't find a metric provider for resource of type: {}".format(resource['type']))
             continue
         log.debug("Sleep 3s to throttle AWS API calls")
         time.sleep(3)
-    sorted_autoscaling_scalers = sorted(autoscaling_scalers, key=lambda x: x.is_resource_cancelled(), reverse=True)
+    filtered_autoscaling_scalers = []
+    for (region, pool), scalers in autoscaling_scalers.items():
+        # if any resource in a pool is cancelled (and it's not scaling up), only look at cancelled ones
+        if any_cancelled_in_pool[(region, pool)] and utilization_errors[(region, pool)] < 0:
+            filtered_autoscaling_scalers += [s for s in scalers if s.is_resource_cancelled()]
+            skipped = [s for s in scalers if s.is_resource_cancelled()]
+            log.info("There are cancelled resources in pool %s, skipping active resources" % pool)
+            log.info("Skipped: %s" % skipped)
+        else:
+            filtered_autoscaling_scalers += scalers
+    sorted_autoscaling_scalers = sorted(
```

Comment:
Would it be easy to pull this out into a separate function that maybe returns `sorted_autoscaling_scalers`? If it's a real pain feel free to ignore me. I just worry about how big some of these methods have got already.

Reply:
It would probably make the tests a bit clearer too